### DISCUSSION TITLE
Implict Operators, Generics, and Nullable

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5972] [Q&A]
I have this code:
```
    public record struct Optional<T> {
        
        public bool IsPresent { get; init; }

        public T? Value { get; init; }

        public static implicit operator Optional<T>(T? Value) {
            // ...
        }
}
```

Which I can use like this:
```
 public class ActivityData {
        public Optional<long> Calendar_Entry_Id { get; set; } = new();
   }

//...

var C = new ActivityData(){ Calendar_Entry_Id = 23 };
```

I would like to be able to run a different code if a null/nullable value is used.  For example:
```
var C = new ActivityData(){ Calendar_Entry_Id = default(long?) };
```
Specifically, in that situation, I want to create an Optional<long> with IsPresent = false instead of an Optional<long?>.

What is the right way to do this?

**Comment by CyrusNajmabadi** [comment]
can't you just check if the value is null here:

```c#
        public static implicit operator Optional<T>(T? Value) {
            return value is null ? default : new { Value = value, IsPresent = true };
        }
```

?

**Reply by TonyValenti** [reply]
@CyrusNajmabadi I cant because the ```Optional<long?>``` operator is called instead of the ```Optional<long>``` operator.

**Reply by CyrusNajmabadi** [reply]
Oh . I see what you're trying to do.  Yes, i don't think this would be possible here (without contraining T which i don't think is your desired intent).

**Comment by jnm2** [comment]
The way I do this is `Optional.FromNullable(expr)` which can have separate `where T : class` and `where T : struct` overloads.

```cs
public static class Optional
{
    public static Optional<T> FromNullable<T>(T? nullableValue) where T : struct
    {
        return nullableValue is { } value ? Optional.Some(value) : Optional.None<T>();
    }

    public static Optional<T> FromNullable<T>(T? nullableValue) where T : class
    {
        return nullableValue is { } value ? Optional.Some(value) : Optional.None<T>();
    }
    
    // Some<T>(T value) and None<T>() are also defined here 
}
```

**Reply by jnm2** [reply]
I don't like implicit operators because then `Optional<object>` can accidentally get boxed and become the `object` inside another `Optional<object>`. I like to see `Optional.Some(...)` around a thing if it's gaining a 'some' semantic.
